File: pyagentspec/src/pyagentspec/adapters/_agentspecloader.py
# Copyright © 2026 Oracle and/or its affiliates.
#
# This software is under the Apache License 2.0
# (LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0) or Universal Permissive License
# (UPL) 1.0 (LICENSE-UPL or https://oss.oracle.com/licenses/upl), at your option.

"""Provide a framework-agnostic loader for Agent Spec configurations."""

import logging
from abc import ABC, abstractmethod
from typing import Any, Callable, Dict, List, Optional, Protocol, TypeAlias, Union, cast, overload

from pyagentspec.component import Component as AgentSpecComponent
from pyagentspec.serialization import AgentSpecDeserializer, ComponentDeserializationPlugin

_RuntimeComponentT: TypeAlias = Any
_RuntimeRegistryT: TypeAlias = Dict[str, Any]


logger = logging.getLogger(__name__)


class AgentSpecToRuntimeConverter(Protocol):
    """Protocol for adapter-specific Agent Spec -> Runtime converters used by loaders."""

    def convert(
        self,
        agentspec_component: AgentSpecComponent,
        tool_registry: Dict[str, Any],
        **kwargs: Any,
    ) -> _RuntimeComponentT:
        """Convert an Agent Spec component into a runtime component."""


class RuntimeToAgentSpecConverter(Protocol):
    """Protocol for adapter-specific Runtime -> Agent Spec converters used by loaders."""

    def convert(
        self,
        runtime_component: _RuntimeComponentT,
        referenced_objects: Optional[Dict[str, AgentSpecComponent]] = None,
        **kwargs: Any,
    ) -> AgentSpecComponent:
        """Convert a runtime component into an Agent Spec component."""


class AdapterAgnosticAgentSpecLoader(ABC):
    """Convert serialized Agent Spec into adapter runtime components.

    This base class centralizes plugin-aware deserialization and support for
    disaggregated components (``import_only_referenced_components``).

    Subclasses supply a converter (typically an ``AgentSpecTo*Converter``) and may
    optionally provide a ``component_registry_converter`` to translate adapter
    runtime registries back to Agent Spec registries for reference resolution.

    Parameters
    ----------
    converter_factory:
        Callable that returns the adapter-specific converter used to convert Agent
        Spec components into runtime components.
    plugins:
        Optional list of Agent Spec deserialization plugins. If omitted, builtin
        plugins compatible with the latest supported Agent Spec version are used.
    component_registry_converter:
        Optional callable to convert a runtime components registry into an Agent
        Spec registry (mapping ids to Agent Spec components/values) so that
        references can be resolved during deserialization.
    """

    def __init__(
        self,
        tool_registry: Optional[Dict[str, Any]] = None,
        plugins: Optional[List[ComponentDeserializationPlugin]] = None,
    ) -> None:
        self.plugins = plugins
        self.tool_registry = tool_registry or {}

    @property
    @abstractmethod
    def agentspec_to_runtime_converter(self) -> AgentSpecToRuntimeConverter:
        """Instance of runtime converter used to convert runtime components."""

    @property
    @abstractmethod
    def runtime_to_agentspec_converter(self) -> RuntimeToAgentSpecConverter:
        """Instance of runtime converter used to convert runtime components."""

    @overload
    def load_yaml(self, serialized_assistant: str) -> _RuntimeComponentT: ...

    @overload
    def load_yaml(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT],
    ) -> _RuntimeComponentT: ...

    @overload
    def load_yaml(
        self,
        serialized_assistant: str,
        *,
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    @overload
    def load_yaml(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT],
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    def load_yaml(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT] = None,
        import_only_referenced_components: bool = False,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]:
        """
        Transform the given Agent Spec YAML into runtime components, with support for disaggregated configurations.

        Parameters
        ----------
        serialized_assistant:
            Serialized Agent Spec configuration.
        components_registry:
            Optional registry mapping ids to runtime components/values. The loader will
            convert these back to Agent Spec components/values internally to resolve references during deserialization.
            If the conversion fails, the given value will be used without conversion.
        import_only_referenced_components:
            When ``True``, loads only the referenced/disaggregated components and returns a
            dictionary mapping component id to runtime components/values.
            These can be used as the ``components_registry`` when loading the main configuration.
            When ``False``, loads the main component and returns the runtime component.

        Returns
        -------
        If ``import_only_referenced_components`` is ``False``

        _RuntimeComponentT
            The runtime component.

        If ``import_only_referenced_components`` is ``True``

        Dict[str, _RuntimeComponentT]
            A dictionary containing the converted referenced components.
        """
        return self._load(
            loader="yaml",
            serialized_assistant=serialized_assistant,
            components_registry=components_registry,
            import_only_referenced_components=import_only_referenced_components,
        )

    @overload
    def load_json(self, serialized_assistant: str) -> _RuntimeComponentT: ...

    @overload
    def load_json(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT],
    ) -> _RuntimeComponentT: ...

    @overload
    def load_json(
        self,
        serialized_assistant: str,
        *,
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    @overload
    def load_json(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT],
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    def load_json(
        self,
        serialized_assistant: str,
        components_registry: Optional[_RuntimeRegistryT] = None,
        import_only_referenced_components: bool = False,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]:
        """
        Transform the given Agent Spec JSON into runtime components, with support for disaggregated configurations.

        Parameters
        ----------
        serialized_assistant:
            Serialized Agent Spec configuration.
        components_registry:
            Optional registry mapping ids to values, runtime components, and Agent Spec components.
            The loader will convert the latter to the respective runtime components during loading.
            If the conversion fails, the given value will be used without conversion.
        import_only_referenced_components:
            When ``True``, loads only the referenced/disaggregated components and returns a
            dictionary mapping component id to runtime components and values.
            These can be used as the ``components_registry`` when loading the main configuration.
            When ``False``, loads the main component and returns the runtime component.

        Returns
        -------
        If ``import_only_referenced_components`` is ``False``

        _RuntimeComponentT
            The runtime component.

        If ``import_only_referenced_components`` is ``True``

        Dict[str, _RuntimeComponentT]
            A dictionary containing the converted referenced components.
        """
        return self._load(
            loader="json",
            serialized_assistant=serialized_assistant,
            components_registry=components_registry,
            import_only_referenced_components=import_only_referenced_components,
        )

    @overload
    def load_dict(self, serialized_assistant: Dict[str, Any]) -> _RuntimeComponentT: ...

    @overload
    def load_dict(
        self,
        serialized_assistant: Dict[str, Any],
        components_registry: Optional[_RuntimeRegistryT],
    ) -> _RuntimeComponentT: ...

    @overload
    def load_dict(
        self,
        serialized_assistant: Dict[str, Any],
        *,
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    @overload
    def load_dict(
        self,
        serialized_assistant: Dict[str, Any],
        components_registry: Optional[_RuntimeRegistryT],
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]: ...

    def load_dict(
        self,
        serialized_assistant: Dict[str, Any],
        components_registry: Optional[_RuntimeRegistryT] = None,
        import_only_referenced_components: bool = False,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]:
        """
        Transform the given Agent Spec dictionary into runtime components, with support for disaggregated configurations.

        Parameters
        ----------
        serialized_assistant:
            Serialized Agent Spec configuration as a dictionary.
        components_registry:
            Optional registry mapping ids to values, runtime components, and Agent Spec components.
            The loader will convert the latter to the respective runtime components during loading.
            If the conversion fails, the given value will be used without conversion.
        import_only_referenced_components:
            When ``True``, loads only the referenced/disaggregated components and returns a
            dictionary mapping component id to runtime components and values.
            These can be used as the ``components_registry`` when loading the main configuration.
            When ``False``, loads the main component and returns the runtime component.

        Returns
        -------
        If ``import_only_referenced_components`` is ``False``

        _RuntimeComponentT
            The runtime component.

        If ``import_only_referenced_components`` is ``True``

        Dict[str, _RuntimeComponentT]
            A dictionary containing the converted referenced components.
        """
        return self._load(
            loader="dict",
            serialized_assistant=serialized_assistant,
            components_registry=components_registry,
            import_only_referenced_components=import_only_referenced_components,
        )

    def load_component(self, agentspec_component: AgentSpecComponent) -> _RuntimeComponentT:
        """Convert a PyAgentSpec component into a runtime component.

        Subclasses may override this method to pass adapter-specific parameters
        into their converter (e.g., tool registries).
        """
        return self.agentspec_to_runtime_converter.convert(
            agentspec_component, tool_registry=self.tool_registry
        )

    def _convert_component_registry(
        self, runtime_component_registry: _RuntimeRegistryT
    ) -> Dict[str, Union[AgentSpecComponent, Any]]:
        # We try to convert the value we have, if it is not supported for conversion, we just keep the value as-is
        exporter = self.runtime_to_agentspec_converter
        converted_registry: Dict[str, Union[AgentSpecComponent, Any]] = {}
        for custom_id, runtime_component_or_value in runtime_component_registry.items():
            try:
                converted_registry[custom_id] = exporter.convert(runtime_component_or_value)
            except Exception as e:
                logger.warning(
                    "Failed to convert runtime component %s with exception `%s`. Fallback to given value.",
                    custom_id,
                    e,
                )
                converted_registry[custom_id] = runtime_component_or_value
        return converted_registry

    def _load(
        self,
        loader: str,
        serialized_assistant: Union[str, Dict[str, Any]],
        components_registry: Optional[_RuntimeRegistryT],
        import_only_referenced_components: bool,
    ) -> Union[_RuntimeComponentT, Dict[str, _RuntimeComponentT]]:

        deserializer = AgentSpecDeserializer(plugins=self.plugins)
        deserializer_func: Callable[..., Any]
        if loader == "yaml":
            deserializer_func = deserializer.from_yaml
        elif loader == "json":
            deserializer_func = deserializer.from_json
        elif loader == "dict":
            deserializer_func = deserializer.from_dict
        else:
            raise ValueError(
                f"Unsupported loader type: `{loader}`. Expected `dict`, `json`, or `yaml`."
            )

        converted_registry = (
            self._convert_component_registry(components_registry)
            if components_registry is not None
            else None
        )
        if import_only_referenced_components:
            # Loading the disaggregated components
            agentspec_referenced_components = deserializer_func(
                serialized_assistant,
                components_registry=converted_registry,
                import_only_referenced_components=True,
            )
            referenced_components_dict = cast(
                Dict[str, AgentSpecComponent], agentspec_referenced_components
            )
            return {
                component_id: self.load_component(agentspec_component_)
                for component_id, agentspec_component_ in referenced_components_dict.items()
            }

        agentspec_component = cast(
            AgentSpecComponent,
            deserializer_func(
                serialized_assistant,
                components_registry=converted_registry,
                import_only_referenced_components=False,
            ),
        )
        return self.load_component(agentspec_component)

File: tsagentspec/src/adapters/common/agentspec-loader.ts
/**
 * Framework-agnostic loader for Agent Spec configurations. Port of
 * `pyagentspec.adapters._agentspecloader.AdapterAgnosticAgentSpecLoader`.
 *
 * This base class centralizes plugin-aware deserialization, the component
 * load policy, and support for disaggregated components
 * (`importOnlyReferencedComponents`). Subclasses supply the two converters.
 *
 * Divergence from Python: the TS `AgentSpecDeserializer` has no
 * allowed/blocked component parameters, so the policy is enforced on the
 * deserialized component tree (`validateComponentTree`) before conversion,
 * not at parse time.
 */
import type { ComponentBase } from "../../component.js";
import { AgentSpecDeserializer } from "../../serialization/index.js";
import type {
  ComponentDeserializationPlugin,
  ComponentsRegistry,
} from "../../serialization/index.js";
import {
  ComponentLoadPolicy,
  type ComponentPolicyInput,
} from "./component-policy.js";
import type {
  AgentSpecToRuntimeConverter,
  RuntimeToAgentSpecConverter,
} from "./converters.js";

/** Components blocked by default: stdio MCP transports run local processes. */
const DEFAULT_BLOCKED_COMPONENTS: readonly string[] = ["StdioTransport"];

/** Constructor options for the adapter-agnostic loader base. */
export interface AdapterAgnosticAgentSpecLoaderOptions {
  /** Registry mapping tool names to runtime implementations. */
  toolRegistry?: Record<string, unknown>;
  /** Deserialization plugins; builtins are used when omitted. */
  plugins?: ComponentDeserializationPlugin[];
  /**
   * Component type names allowed to load. When omitted, all component types
   * are allowed unless blocked.
   */
  allowedComponents?: ComponentPolicyInput;
  /**
   * Component type names blocked from loading. When omitted, `StdioTransport`
   * is blocked by default.
   */
  blockedComponents?: ComponentPolicyInput;
}

/** Per-call options for the load methods. */
export interface LoadOptions {
  /**
   * Registry mapping ids to runtime components/values. Entries are converted
   * back to Agent Spec components to resolve references during
   * deserialization; if a conversion fails, the given value is used as-is.
   */
  componentsRegistry?: Record<string, unknown>;
  /**
   * When true, load only the referenced/disaggregated components and return a
   * dictionary mapping component id to runtime components/values. These can
   * be used as the `componentsRegistry` when loading the main configuration.
   */
  importOnlyReferencedComponents?: boolean;
}

/** Convert serialized Agent Spec into adapter runtime components. */
export abstract class AdapterAgnosticAgentSpecLoader {
  /** Registry mapping tool names to runtime implementations. */
  readonly toolRegistry: Record<string, unknown>;
  /** Deserialization plugins passed to the deserializer (builtins if unset). */
  readonly plugins?: ComponentDeserializationPlugin[];
  /** The allow/block policy applied to every loaded component tree. */
  readonly componentLoadPolicy: ComponentLoadPolicy;
  /** Normalized allow-list entries, or undefined when no allow list is set. */
  readonly allowedComponents?: readonly string[];
  /** Normalized block-list entries. */
  readonly blockedComponents: readonly string[];

  constructor(options?: AdapterAgnosticAgentSpecLoaderOptions) {
    const opts = options ?? {};
    this.plugins = opts.plugins;
    this.toolRegistry = opts.toolRegistry ?? {};
    this.componentLoadPolicy = new ComponentLoadPolicy(
      opts.allowedComponents,
      opts.blockedComponents ?? DEFAULT_BLOCKED_COMPONENTS,
    );
    this.allowedComponents = this.componentLoadPolicy.allowedComponents;
    this.blockedComponents = this.componentLoadPolicy.blockedComponents;
  }

  /** Converter used to convert Agent Spec components to runtime components. */
  abstract get agentspecToRuntimeConverter(): AgentSpecToRuntimeConverter;

  /** Converter used to convert runtime components to Agent Spec components. */
  abstract get runtimeToAgentSpecConverter(): RuntimeToAgentSpecConverter;

  /**
   * Transform the given Agent Spec YAML into runtime components, with support
   * for disaggregated configurations.
   */
  async loadYaml(
    serializedAssistant: string,
    options?: LoadOptions,
  ): Promise<unknown> {
    return this._load("yaml", serializedAssistant, options);
  }

  /**
   * Transform the given Agent Spec JSON into runtime components, with support
   * for disaggregated configurations.
   */
  async loadJson(
    serializedAssistant: string,
    options?: LoadOptions,
  ): Promise<unknown> {
    return this._load("json", serializedAssistant, options);
  }

  /**
   * Transform the given Agent Spec dictionary into runtime components, with
   * support for disaggregated configurations.
   */
  async loadDict(
    serializedAssistant: Record<string, unknown>,
    options?: LoadOptions,
  ): Promise<unknown> {
    return this._load("dict", serializedAssistant, options);
  }

  /**
   * Convert an Agent Spec component into a runtime component after validating
   * it against the component load policy.
   *
   * Subclasses may override this method to pass adapter-specific parameters
   * into their converter (e.g., checkpointers).
   */
  async loadComponent(agentspecComponent: ComponentBase): Promise<unknown> {
    this.componentLoadPolicy.validateComponentTree(agentspecComponent);
    return this.agentspecToRuntimeConverter.convert(
      agentspecComponent,
      this.toolRegistry,
    );
  }

  /**
   * Convert a runtime components registry into an Agent Spec registry so that
   * references can be resolved during deserialization. Values that fail to
   * convert are kept as-is with a warning (mirrors Python).
   */
  protected _convertComponentRegistry(
    runtimeComponentRegistry: Record<string, unknown>,
  ): ComponentsRegistry {
    const converter = this.runtimeToAgentSpecConverter;
    const convertedRegistry: ComponentsRegistry = new Map();
    for (const [customId, runtimeComponentOrValue] of Object.entries(
      runtimeComponentRegistry,
    )) {
      try {
        convertedRegistry.set(
          customId,
          converter.convert(runtimeComponentOrValue),
        );
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        console.warn(
          `Failed to convert runtime component ${customId} with exception \`${message}\`. Fallback to given value.`,
        );
        // Mirrors Python: unconvertible registry values are used as-is when
        // resolving references.
        convertedRegistry.set(customId, runtimeComponentOrValue as ComponentBase);
      }
    }
    return convertedRegistry;
  }

  /** Common implementation of the load methods. */
  protected async _load(
    loader: "yaml" | "json" | "dict",
    serializedAssistant: string | Record<string, unknown>,
    options?: LoadOptions,
  ): Promise<unknown> {
    const deserializer = new AgentSpecDeserializer(this.plugins);
    let deserialize: (deserializeOptions: {
      componentsRegistry?: ComponentsRegistry;
      importOnlyReferencedComponents?: boolean;
    }) => ComponentBase | Record<string, ComponentBase>;
    if (loader === "yaml") {
      deserialize = (deserializeOptions) =>
        deserializer.fromYaml(serializedAssistant as string, deserializeOptions);
    } else if (loader === "json") {
      deserialize = (deserializeOptions) =>
        deserializer.fromJson(serializedAssistant as string, deserializeOptions);
    } else if (loader === "dict") {
      // The TS AgentSpecDeserializer has no public dict entry point;
      // round-trip through JSON.
      const json = JSON.stringify(serializedAssistant);
      deserialize = (deserializeOptions) =>
        deserializer.fromJson(json, deserializeOptions);
    } else {
      throw new Error(
        `Unsupported loader type: \`${String(loader)}\`. Expected \`dict\`, \`json\`, or \`yaml\`.`,
      );
    }

    const convertedRegistry =
      options?.componentsRegistry !== undefined
        ? this._convertComponentRegistry(options.componentsRegistry)
        : undefined;

    if (options?.importOnlyReferencedComponents) {
      // Loading the disaggregated components
      const referencedComponentsDict = deserialize({
        componentsRegistry: convertedRegistry,
        importOnlyReferencedComponents: true,
      }) as Record<string, ComponentBase>;
      const runtimeComponents: Record<string, unknown> = {};
      for (const [componentId, agentspecComponent] of Object.entries(
        referencedComponentsDict,
      )) {
        runtimeComponents[componentId] =
          await this.loadComponent(agentspecComponent);
      }
      return runtimeComponents;
    }

    const agentspecComponent = deserialize({
      componentsRegistry: convertedRegistry,
      importOnlyReferencedComponents: false,
    }) as ComponentBase;
    return this.loadComponent(agentspecComponent);
  }
}
